Add support for AWS IAM authentication using boto3

kobuskc opened the pull request:
Using the get_credentials definition, we can get credentials from the local environment.

Definition: http://boto3.readthedocs.io/en/latest/reference/core/session.html#boto3.session.Session.get_credentials

Order in which credentials is searched: http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#config-settings-and-precedence

untergeek commented on curator/utils.py:
```diff
@@ -745,36 +751,48 @@ def get_client(**kwargs):
-            else kwargs['aws_region']
+        kwargs['aws_sign_request'] = False if not 'aws_sign_request' in kwargs \
+            else kwargs['aws_sign_request']
+        if kwargs['aws_sign_request']:
```
This is going to need to be more complex.

Not only does it have to `try`/`except` on the import module, it has to `try`/`except` on the `session.Session()` (because it will raise an exception if it is unable, and we should control how it acts if it does, and log good messages).  If it is unable to get the session, then it would then rely on the original `aws_key` and `aws_secret_key`, which should be set first.

The contents of the `else` block should precede the `if kwargs['aws_sign_request']` statement as a pre-population thing, even if they end up being `False` because they were not specified.  The `boto3` stuff will overwrite the necessary values if it exists, so it is completely okay to have `kwargs['aws_key']` and `kwargs['aws_secret_key']` set first.

Does this make sense?

kobuskc replied:
This makes sense. I have implemented your suggestions, so now credentials retrieved with boto3 will overwrite anything that was in the config file if the user specified both.

untergeek commented on curator/defaults/client_defaults.py:
```diff
@@ -14,7 +14,9 @@ def config_client():
-        Optional('aws_region', default=None): Any(None, str, unicode),
+        Optional('aws_token', default=None): Any(None, str, unicode),
+        Optional('aws_sign_request', default=False): Boolean(),
+        Optional('aws_region', default='us-east-1'): Any(None, str, unicode),
```
Do you think it's wise to set the default to `us-east-1`, rather than just leave it kind of required to be set if you're using AWS credentials?

kobuskc replied:
Makes sense. I have removed the default.

kobuskc commented:
Not sure why the Travis build failed. This is my test output:

----------------------------------------------------------------------
XML: /curator/nosetests.xml
Name                                  Stmts   Miss  Cover
---------------------------------------------------------
curator/__init__.py                      11      0   100%
curator/_version.py                       1      0   100%
curator/actions.py                      870    113    87%
curator/cli.py                          113      0   100%
curator/config_utils.py                  34      5    85%
curator/defaults/__init__.py              0      0   100%
curator/defaults/client_defaults.py       5      0   100%
curator/defaults/filter_elements.py      66      0   100%
curator/defaults/filtertypes.py          60      0   100%
curator/defaults/option_defaults.py     147      0   100%
curator/defaults/settings.py             32      0   100%
curator/exceptions.py                     9      0   100%
curator/indexlist.py                    369      0   100%
curator/logtools.py                      36      0   100%
curator/repomgrcli.py                    78      1    99%
curator/snapshotlist.py                 186      0   100%
curator/utils.py                        688     11    98%
curator/validators/__init__.py            1      0   100%
curator/validators/actions.py            19      0   100%
curator/validators/config_file.py         4      0   100%
curator/validators/filters.py            32      0   100%
curator/validators/options.py            13      0   100%
curator/validators/schemacheck.py        41      0   100%
---------------------------------------------------------
TOTAL                                  2815    130    95%
----------------------------------------------------------------------
Ran 570 tests in 136.390s

OK (SKIP=4)
Shutting down....
[root@xxxxxxxxxxx curator]# 
[nosetests.xml.gz](https://github.com/elastic/curator/files/1384222/nosetests.xml.gz)

untergeek commented on curator/utils.py:
```diff
@@ -745,15 +751,40 @@ def get_client(**kwargs):
+        # If an attribute doesn't exist, we were not able to retrieve credentials as expected so we can't continue
+        except AttributeError:
+            logger.debug('Unable to locate AWS credentials')
+            exit('Unable to locate AWS credentials')
```
Don't do an `exit` here.  This should simply raise an Exception of some kind, either a `ConfigurationError`, indicating something is misconfigured, or a `CuratorException`, as there isn't an exception in exceptions.py that's more specific.

```
except AttributeError:
    raise CuratorException('Unable to locate AWS credentials, even though "aws_sign_request" was specified')
```

untergeek commented on curator/utils.py:
```diff
@@ -745,15 +751,40 @@ def get_client(**kwargs):
+        # If an attribute doesn't exist, we were not able to retrieve credentials as expected so we can't continue
+        except AttributeError:
+            logger.debug('Unable to locate AWS credentials')
+            raise exceptions.NoCredentialsError
```
`NoCredentialsError` is not a valid member of the [base exceptions class](https://docs.python.org/3/library/exceptions.html).  Is that in the `boto` module?  If so, it would just be `NoCredentialsError`, rather than `exceptions.NoCredentialsError`.

untergeek commented:
Something is still not working with this code.  Client instances are not being created, causing every integration test to fail.

kobuskc commented:
Changed the import of botocore exceptions for clarity and fixed the logic error that prevented Client instances from being created. Local test results:

----------------------------------------------------------------------
XML: /curator/nosetests.xml
Name                                  Stmts   Miss  Cover
---------------------------------------------------------
curator/__init__.py                      11      0   100%
curator/_version.py                       1      0   100%
curator/actions.py                      870    113    87%
curator/cli.py                          113      0   100%
curator/config_utils.py                  34      5    85%
curator/defaults/__init__.py              0      0   100%
curator/defaults/client_defaults.py       5      0   100%
curator/defaults/filter_elements.py      66      0   100%
curator/defaults/filtertypes.py          60      0   100%
curator/defaults/option_defaults.py     147      0   100%
curator/defaults/settings.py             32      0   100%
curator/exceptions.py                     9      0   100%
curator/indexlist.py                    369      0   100%
curator/logtools.py                      36      0   100%
curator/repomgrcli.py                    78      1    99%
curator/snapshotlist.py                 186      0   100%
curator/utils.py                        710     29    96%
curator/validators/__init__.py            1      0   100%
curator/validators/actions.py            19      0   100%
curator/validators/config_file.py         4      0   100%
curator/validators/filters.py            32      0   100%
curator/validators/options.py            13      0   100%
curator/validators/schemacheck.py        41      0   100%
---------------------------------------------------------
TOTAL                                  2837    148    95%
----------------------------------------------------------------------
Ran 570 tests in 136.873s

OK (SKIP=4)
Shutting down....

untergeek commented:
So, is this connecting using `boto` now on your end?

If that part is working, the next step will be to add the proper documentation to [`docs/asciidoc/configuration.asciidoc`](https://github.com/elastic/curator/blob/v5.2.0/docs/asciidoc/configuration.asciidoc).  You may even need to edit existing information with regards to `aws_key` and the other related fields.  As these changes you are adding are not tested by the Travis CI testing, please leave them as such: hand tested.  I do not have the capability to test them, so they are a community addition, and should be properly vetted and tested by end users with the `--dry-run` flag, etc. before using.

kobuskc commented:
Yes, it connects with <boto> as expected. Here's an example test with <--dry-run>:

### With aws_sign_requests: True
xxxxxxxxxx  ~/actions  » curator --dry-run snapaction.yml 
2017-10-19 08:24:21,613 INFO      Preparing Action ID: 1, "snapshot"
**2017-10-19 08:24:21,756 INFO      Found credentials in shared credentials file: ~/.aws/credentials**
2017-10-19 08:24:21,817 INFO      Trying Action ID: 1, "snapshot": Snapshot members prefixed indices older than 1 day (based on index creation_date) with the default snapshot name pattern of 'curator-%Y%m%d%H%M%S'.  Wait for the snapshot to complete.  Do not skip the repository filesystem access check.  Use the other options to create the snapshot.
2017-10-19 08:24:21,875 INFO      DRY-RUN MODE.  No changes will be made.
2017-10-19 08:24:21,875 INFO      DRY-RUN: snapshot: curator-20171018212421 in repository backups with arguments: {'ignore_unavailable': False, 'partial': False, 'indices': 'members', 'include_global_state': True}
2017-10-19 08:24:21,875 INFO      Action ID: 1, "snapshot" completed.
2017-10-19 08:24:21,875 INFO      Job completed. 

### With no aws_sign_requests, and both aws_key and aws_secret_key specified
xxxxxxxxxx  ~/actions  » curator --dry-run snapaction.yml
2017-10-19 08:26:03,700 INFO      Preparing Action ID: 1, "snapshot"
2017-10-19 08:26:03,839 INFO      Trying Action ID: 1, "snapshot": Snapshot members prefixed indices older than 1 day (based on index creation_date) with the default snapshot name pattern of 'curator-%Y%m%d%H%M%S'.  Wait for the snapshot to complete.  Do not skip the repository filesystem access check.  Use the other options to create the snapshot.
2017-10-19 08:26:03,890 INFO      DRY-RUN MODE.  No changes will be made.
2017-10-19 08:26:03,890 INFO      DRY-RUN: snapshot: curator-20171018212603 in repository backups with arguments: {'partial': False, 'ignore_unavailable': False, 'include_global_state': True, 'indices': 'members'}
2017-10-19 08:26:03,890 INFO      Action ID: 1, "snapshot" completed.
2017-10-19 08:26:03,890 INFO      Job completed.

I will update the documentation today.

untergeek commented on docs/asciidoc/configuration.asciidoc:
```diff
@@ -420,6 +426,37 @@ IMPORTANT: You must set your <<hosts,hosts>> to the proper hostname _with_ port.
+WARNING: This feature has not been fully tested and should be considered BETA.
+
+WARNING: This setting will not work unless the `requests-aws4auth` and `boto3` Python
+    modules have been manually installed first.
```
This should probably read, "If installing via `pip`, this setting will not work unless ...".  I install `requests-aws4auth` in the [RPM/DEB](https://github.com/elastic/curator/blob/v5.2.0/unix_packages/build_official_package.sh#L143) packages.  If you want to edit both of the files in `unix_packages` directory to add `boto` to be installed, that would be okay, too.

untergeek commented on docs/asciidoc/configuration.asciidoc:
```diff
@@ -420,6 +426,37 @@ IMPORTANT: You must set your <<hosts,hosts>> to the proper hostname _with_ port.
+
+[source,sh]
+-----------
+aws_sign_request:
```
This should be `False` or `True` here, don't you think?

untergeek commented on docs/asciidoc/configuration.asciidoc:
```diff
@@ -420,6 +426,37 @@ IMPORTANT: You must set your <<hosts,hosts>> to the proper hostname _with_ port.
+-----------
+aws_sign_request:
+-----------
+
```
There should be a line somewhere indicating the default value (which I presume is `False`, right?).

untergeek commented on curator/utils.py:
```diff
@@ -745,36 +751,60 @@ def get_client(**kwargs):
+            kwargs['aws_token'] = credentials.token
+        # If an attribute doesn't exist, we were not able to retrieve credentials as expected so we can't continue
+        except AttributeError:
+            logger.debug('Unable to locate AWS credentials')
```
Hmmm.  I'll let this slide, since we're raising the `NoCredentialsError` anyway, but it seems like this should be `logger.error`, rather than `logger.debug`.

untergeek commented:
Thanks for your work on this.  Hopefully it results in some very happy AWS users.
